# 01 Importing libraries

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
# Create strings for path
path=r'C:\Users\Oksana Stepanova\OneDrive\Документи\Instacart Basket Analysis'

In [4]:
# Import merged Orders-Products dataframe
df_ords_prods_merge=pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [5]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  
0          0         Soda        77              7     9.0   both  
1          1         Soda        77              7     9.0   both  
2          1         Soda        77              7     9.0   both  
3          1         Soda        77              7     9.0   both  
4          1         Soda        77              7     9.0   both

In [6]:
df_ords_prods_merge.shape

(32399732, 15)

In [7]:
# Create a subset of the first 1 mio rows
df=df_ords_prods_merge[:1000000]

In [8]:
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  
0          0         Soda        77              7     9.0   both  
1          1         Soda        77              7     9.0   both  
2          1         Soda        77              7     9.0   both  
3          1         Soda        77              7     9.0   both  
4          1         Soda        77              7     9.0   both

In [9]:
df.shape

(1000000, 15)

In [10]:
df.dtypes

order_id                     int32
user_id                      int32
order_number                 int16
orders_day_of_week            int8
order_hour_of_day             int8
days_since_prior_order     float64
first_order                   bool
product_id                   int32
add_to_cart_order            int16
reordered                     int8
product_name                object
aisle_id                     int16
department_id                 int8
prices                     float64
_merge                    category
dtype: object

In [11]:
df.describe()

order_id         user_id    order_number  orders_day_of_week  \
count  1.000000e+06  1000000.000000  1000000.000000      1000000.000000   
mean   1.710240e+06   101788.547547       18.731714            2.686741   
std    9.873682e+05    59011.894306       18.563841            2.053382   
min    5.000000e+00        1.000000        1.000000            0.000000   
25%    8.555392e+05    50641.000000        5.000000            1.000000   
50%    1.709734e+06   101164.000000       12.000000            2.000000   
75%    2.565079e+06   152132.000000       27.000000            5.000000   
max    3.421068e+06   206209.000000       99.000000            6.000000   

       order_hour_of_day  days_since_prior_order      product_id  \
count     1000000.000000           939915.000000  1000000.000000   
mean           13.272235               10.646367    27119.966748   
std             4.218017                8.467237    15578.819637   
min             0.000000                0.000000      196.000000   
25%            10.000000                5.000000    13176.000000   
50%            13.000000                7.000000    26088.000000   
75%            16.000000               14.000000    47766.000000   
max            23.000000               30.000000    49235.000000   

       add_to_cart_order       reordered        aisle_id   department_id  \
count     1000000.000000  1000000.000000  1000000.000000  1000000.000000   
mean            6.078493        0.769933       44.407396        8.740095   
std             5.695749        0.420876       28.925073        6.419989   
min             1.000000        0.000000       21.000000        4.000000   
25%             2.000000        1.000000       24.000000        4.000000   
50%             4.000000        1.000000       24.000000        4.000000   
75%             8.000000        1.000000       67.000000       16.000000   
max            99.000000        1.000000      121.000000       20.000000   

               prices  
count  1000000.000000  
mean         7.819497  
std          3.641458  
min          1.000000  
25%          6.300000  
50%          9.000000  
75%         10.300000  
max         14.800000

# 03 Using defined function

In [14]:
# Define the function
def price_label(row):
    if row['prices']<=5:
        return 'Low range product'
    elif (row['prices']>5) and (row['prices']<=15):
        return 'Midle range product'
    elif row['prices']>15:
        return 'High range product'
    else: return np.nan

In [15]:
# Apply the function
df['price_range']=df.apply(price_label, axis=1)

C:\Users\Oksana Stepanova\AppData\Local\Temp\ipykernel_316\969816975.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range']=df.apply(price_label, axis=1)


In [16]:
df['price_range'].value_counts(dropna=False)

price_range
Midle range product    756450
Low range product      243550
Name: count, dtype: int64

In [17]:
df['prices'].max()

14.8

In [18]:
df['prices'].describe()

count    1000000.000000
mean           7.819497
std            3.641458
min            1.000000
25%            6.300000
50%            9.000000
75%           10.300000
max           14.800000
Name: prices, dtype: float64

# 04 Using the loc() function

In [20]:
# Create condition for 'prices'>15
df.loc[df['prices']>15, 'price_range_loc']='High range product'

C:\Users\Oksana Stepanova\AppData\Local\Temp\ipykernel_316\986652602.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices']>15, 'price_range_loc']='High range product'


In [22]:
# Create condition for 'prices'<=15 and >5
df.loc[(df['prices']<=15) & (df['prices']>5), 'price_range_loc']='Mid range product'

In [23]:
# Create condition for 'prices'<=5
df.loc[df['prices']<=5, 'price_range_loc']='Low range product'

In [24]:
df['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid range product    756450
Low range product    243550
Name: count, dtype: int64

In [26]:
# Range products on merged Orders-Products dataframe
df_ords_prods_merge.loc[df_ords_prods_merge['prices']>15, 'price_range_loc']='High range product'

In [27]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices']<=15) & (df_ords_prods_merge['prices']>5), 'price_range_loc']='Mid range product'

In [28]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices']<=5, 'price_range_loc']='Low range product'

In [29]:
# Check for values in the new column of product ranging
df_ords_prods_merge['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid range product     21860860
Low range product     10126321
High range product      412551
Name: count, dtype: int64

In [30]:
# Check for data type
df_ords_prods_merge.dtypes

order_id                     int32
user_id                      int32
order_number                 int16
orders_day_of_week            int8
order_hour_of_day             int8
days_since_prior_order     float64
first_order                   bool
product_id                   int32
add_to_cart_order            int16
reordered                     int8
product_name                object
aisle_id                     int16
department_id                 int8
prices                     float64
_merge                    category
price_range_loc             object
dtype: object

In [31]:
# Check for frequency in oreders_day_of_week column
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6203329
1    5659298
6    4495887
2    4213105
5    4205076
3    3839865
4    3783172
Name: count, dtype: int64

# 05 Using For-Loops

In [35]:
# Create a loop to range the busiest day of the week
result=[]
for value in df_ords_prods_merge['orders_day_of_week']:
    if value==0:
        result.append("Busiest day")
    elif value==4:
            result.append('Least busy')
    else:
                result.append('Regular busy')

In [36]:
result

['Regular busy',
 'Regular busy',
 'Regular busy',
 'Least busy',
 'Least busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Least busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Least busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Least busy',
 'Least busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Least busy',
 'Regular busy',
 'Busiest day',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular bu

In [37]:
# Create a new column and set it equal to result
df_ords_prods_merge['busiest_day']=result

In [40]:
# Check the values in the new column 
df_ords_prods_merge['busiest_day'].value_counts(dropna=False)

busiest_day
Regular busy    22413231
Busiest day      6203329
Least busy       3783172
Name: count, dtype: int64

# 06 Task 4.7. Step 2

In [59]:
# Create a loop to range two busiest days and two slowest days of the week
result2=[]
for value in df_ords_prods_merge['orders_day_of_week']:
    if value==0 or value==1:
        result2.append('Busiest days')
    elif value==3 or value==4:
            result2.append('Slowest days')
    else:
                result2.append('Regular busy')

In [56]:
result2

['Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regular busy',
 'Busiest days',
 'Regular busy',
 'Regular busy',
 'Busiest days',
 'Busiest days',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regular busy',
 'Regular busy',
 'Busiest days',
 'Regular busy',
 'Regular busy',
 'Busiest days',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Busiest days',
 'Busiest days',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Regular busy',
 'Busiest days',
 'Busiest days',
 'Regular busy',
 'Busiest days',
 'Busiest days',
 'Regular busy',
 'Regular busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regular busy',
 'Regular busy

In [60]:
# Create a new column and set it equal to result2
df_ords_prods_merge['busiest_slowest_days']=result2

In [61]:
# Check the values in the new column of two busiest and two slowest days of the week 
df_ords_prods_merge['busiest_slowest_days'].value_counts(dropna=False)

busiest_slowest_days
Regular busy    12914068
Busiest days    11862627
Slowest days     7623037
Name: count, dtype: int64

# 07 Task 4.7. Step 3

In [63]:
# Check for the output of Orders-Products dataframe with a new column of busiest and slowest days of the week
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

     price_range_loc   busiest_day busiest_slowest_days  
0  Mid range product  Regular busy         Regular busy  
1  Mid range product  Regular busy         Slowest days  
2  Mid range product  Regular busy         Slowest days  
3  Mid range product    Least busy         Slowest days  
4  Mid range product    Least busy         Slowest days

In [64]:
df_ords_prods_merge.shape

(32399732, 18)

In [65]:
df_ords_prods_merge['busiest_slowest_days'].isnull().sum()

0

In [66]:
df_ords_prods_merge['busiest_slowest_days'].shape

(32399732,)

# In the column ‘busiest_slowest_days’, there are no missing values. The count of ‘Busiest days’ values (11,862,627) matches the total of ‘0’ and ‘1’ values in ‘orders_day_of_week’ column. We can note the same for the ‘Slowest days’ , the count in the new column (7,323,037) corresponds to the total of ‘3’ and ‘4’ in the in ‘orders_day_of_week’ column. After adding a new column, the total number of rows in the dataframe is the same and equal to 32,399,732.  

# 08 Task 4.7. Step 4. Creating a time labeled column for busiest hours of the day 

In [70]:
# Check for frequency in oreder_hour_of_day column
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna=False)

order_hour_of_day
10    2761333
11    2735694
14    2688728
15    2661718
13    2660570
12    2618104
16    2534744
9     2453842
17    2087273
8     1717863
18    1636226
19    1258076
20     976000
7      890923
21     795528
22     634159
23     402272
6      290450
0      218742
1      115683
5       87944
2       69360
4       53232
3       51268
Name: count, dtype: int64

# The first eight values in the frequency list will be labeled as ‘Most orders’ (9, 10, 11, 12, 13, 14, 15, 16) and the last eight hours will be labeled as ‘Fewest orders’ accordingly (0, 1, 2, 3, 4, 5, 6, 23). The eight values in the middle of the list will be added as ‘Average orders’ (7, 8, 17, 18, 19, 20, 21, 22)

In [74]:
# Create a loop to range the busiest and slowest hours of the day
result4=[]
for value in df_ords_prods_merge['order_hour_of_day']:
    if value in [9, 10, 11, 12, 13, 14, 15, 16]:
        result4.append("Most orders")
    elif value in [0, 1, 2, 3, 4, 5, 6, 23]:
            result4.append("Fewest orders")
    else:
                result4.append("Average orders")

In [75]:
result4

['Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most ord

In [76]:
# Create a new column 'busiest_period_of_day' and set it equal to result4
df_ords_prods_merge['busiest_period_of_day']=result4

In [78]:
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna=False)

busiest_period_of_day
Most orders       21114733
Average orders     9996048
Fewest orders      1288951
Name: count, dtype: int64

In [79]:
df_ords_prods_merge.shape

(32399732, 19)

In [80]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

     price_range_loc   busiest_day busiest_slowest_days busiest_period_of_day  
0  Mid range product  Regular busy         Regular busy        Average orders  
1  Mid range product  Regular busy         Slowest days        Average orders  
2  Mid range product  Regular busy         Slowest days           Most orders  
3  Mid range product    Least busy         Slowest days        Average orders  
4  Mid range product    Least busy         Slowest days           Most orders

In [81]:
# Drop column '_merge'
df_ords_prods_merge_drop=df_ords_prods_merge.drop(columns=['_merge'])

In [82]:
# Check for the number of rows and columns and the representation of Orders-Product dataframe after removing the column
df_ords_prods_merge_drop.shape

(32399732, 18)

In [83]:
df_ords_prods_merge_drop.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered product_name  aisle_id  department_id  prices    price_range_loc  \
0          0         Soda        77              7     9.0  Mid range product   
1          1         Soda        77              7     9.0  Mid range product   
2          1         Soda        77              7     9.0  Mid range product   
3          1         Soda        77              7     9.0  Mid range product   
4          1         Soda        77              7     9.0  Mid range product   

    busiest_day busiest_slowest_days busiest_period_of_day  
0  Regular busy         Regular busy        Average orders  
1  Regular busy         Slowest days        Average orders  
2  Regular busy         Slowest days           Most orders  
3    Least busy         Slowest days        Average orders  
4    Least busy         Slowest days           Most orders

# 08 Exporting data

In [85]:
df_ords_prods_merge_drop.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_new_vars.pkl'))